roachtest: scaledata/distributed_semaphore/nodes=3 failed

SHA: https://github.com/cockroachdb/cockroach/commits/766a53defc27940d41ae8987792ecd79e76d2b65

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=scaledata/distributed_semaphore/nodes=3 PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1090948&tab=buildLog

```
The test failed on master:
	test.go:1101: test timed out (20m0s)
	test.go:696,cluster.go:1511,scaledata.go:126,scaledata.go:53: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1090948-scaledata-distributed-semaphore-nodes-3:4 -- ./distributed_semaphore  --duration_secs=600 --num_workers=16 --cockroach_ip_addresses_csv='10.128.0.154:26257,10.128.0.124:26257,10.128.0.155:26257'  returned:
		stderr:
		13 12:19:43.818029251 +0000 UTC m=+574.303702067, now = 2019-01-13 12:19:58.760941404 +0000 UTC m=+589.246614276, took 14.942912209s
		2019/01/13 12:19:58 Aborting Retries because this error of type *crdb.AmbiguousCommitError is not retryable : pq: result is ambiguous (error=initial connection heartbeat failed: rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = "transport: Error while dialing dial tcp 10.128.0.124:26257: connect: connection refused" [exhausted])
		2019/01/13 12:19:58 RobustDB.RandomDB chose DB at index 0
		2019/01/13 12:19:58 [/Users/nathan/Go/src/github.com/scaledata/rksql/src/go/src/rubrik/sqlapp/tx.go:29 pq: restart transaction: TransactionRetryWithProtoRefreshError: TransactionRetryError: retry txn (RETRY_ASYNC_WRITE_FAILURE): "sql txn" id=4ed1fae5 key=/Table/54/1/417173639025852417/0 rw=true pri=0.00407482 stat=PENDING epo=0 ts=1547381994.115138047,1 orig=1547381983.836350074,0 max=1547381984.336350074,0 wto=true seq=2]
		
		stdout:
		e timestamp 1547381409.525352672,1 within uncertainty interval `t <= 1547381409.538741774,0`; observed timestamps: [{1 1547381409.538741774,0} {2 1547381409.524794522,0}]
		pq: restart transaction: TransactionRetryWithProtoRefreshError: WriteTooOldError: write at timestamp 1547381409.540432066,0 too old; wrote at 1547381409.540432066,2
		pq: restart transaction: TransactionRetryWithProtoRefreshError: ReadWithinUncertaintyIntervalError: read at time 1547381409.523763080,0 encountered previous write with future timestamp 1547381409.540432066,1 within uncertainty interval `t <= 1547381409.603062510,0`; observed timestamps: [{1 1547381409.603062510,0} {3 1547381409.523763080,0}]
		pq: restart transaction: TransactionRetryWithProtoRefreshError: ReadWithinUncertaintyIntervalError: read at time 1547381409.524120255,0 encountered previous write with future timestamp 1547381409.540432066,1 within uncertainty interval `t <= 1547381409.603154178,0`; observed timestamps: [{1 1547381409.603154178,0} {3 1547381409.524120255,0}]
		: signal: killed

```

Looks like this tickled a zig zag bug:

```
E190320 12:17:53.060953 523 util/log/crash_reporting.go:216  [n3,client=10.142.0.179:55746,user=root] a panic has occurred!
panic: non-values node passed as fixed value to zigzag join [recovered]
	panic: non-values node passed as fixed value to zigzag join [recovered]
	panic: panic while executing 1 statements: SELECT * FROM _ WHERE (_._ IS _) AND (((_ = $1) AND (_ = $2))); caused by non-values node passed as fixed value to zigzag join

goroutine 565 [running]:
panic(0x3103ee0, 0xc00644cab0)
	/usr/local/go/src/runtime/panic.go:556 +0x2cb fp=0xc006433408 sp=0xc006433378 pc=0x72b49b
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc005efd100, 0x39fc4a0, 0xc0062e5c80, 0x2d97f80, 0x39b5110)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:714 +0x36d fp=0xc0064334c8 sp=0xc006433408 pc=0x1fadaad
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1(0xc005efd100, 0x39fc4a0, 0xc0062e5c80)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:426 +0x61 fp=0xc006433500 sp=0xc0064334c8 pc=0x213aa11
runtime.call32(0x0, 0x33f85f0, 0xc005377700, 0x1800000018)
	/usr/local/go/src/runtime/asm_amd64.s:522 +0x3b fp=0xc006433530 sp=0xc006433500 pc=0x75a06b
panic(0x2d97f80, 0x39b5110)
	/usr/local/go/src/runtime/panic.go:513 +0x1b9 fp=0xc0064335c0 sp=0xc006433530 pc=0x72b389
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).Build.func1(0xc0064349a8)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/builder.go:66 +0x9b fp=0xc0064335e8 sp=0xc0064335c0 pc=0x1bcb19b
runtime.call32(0x0, 0x33f9038, 0xc000094a70, 0x800000008)
	/usr/local/go/src/runtime/asm_amd64.s:522 +0x3b fp=0xc006433618 sp=0xc0064335e8 pc=0x75a06b
panic(0x2d97f80, 0x39b5110)
	/usr/local/go/src/runtime/panic.go:513 +0x1b9 fp=0xc0064336a8 sp=0xc006433618 pc=0x72b389
github.com/cockroachdb/cockroach/pkg/sql.(*execFactory).ConstructZigzagJoin(0xc005b09a88, 0x3a3c160, 0xc0064eb040, 0x3a2aa80, 0xc005da5d28, 0x3a3c160, 0xc0064eb040, 0x3a2aa80, 0xc005da5d90, 0xc0064f0bb8, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt_exec_factory.go:736 +0x8ea fp=0xc006433758 sp=0xc0064336a8 pc=0x206d37a
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildZigzagJoin(0xc0064462a0, 0xc0063f71f0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc006433ba0, 0x1aef637)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational_builder.go:1235 +0xe46 fp=0xc006433b10 sp=0xc006433758 pc=0x1bc0626
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildRelational(0xc0064462a0, 0x3a36200, 0xc0063f71f0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational_builder.go:190 +0x732 fp=0xc006434058 sp=0xc006433b10 pc=0x1bb5022
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildLookupJoin(0xc0064462a0, 0xc0064e0a00, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc0064343c0, 0xc0064343c8)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational_builder.go:1128 +0x97 fp=0xc006434328 sp=0xc006434058 pc=0x1bbedf7
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildRelational(0xc0064462a0, 0x3a353a0, 0xc0064e0a00, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc000094a40, 0x70b01f)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational_builder.go:187 +0xe31 fp=0xc006434870 sp=0xc006434328 pc=0x1bb5721
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).build(0xc0064462a0, 0x3a0a4e0, 0xc0064e0a00, 0x0, 0xc006434980, 0x70b8a8, 0x60)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/builder.go:83 +0xab fp=0xc006434930 sp=0xc006434870 pc=0x1bb36fb
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).Build(0xc0064462a0, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/builder.go:71 +0x78 fp=0xc006434990 sp=0xc006434930 pc=0x1bb3578
github.com/cockroachdb/cockroach/pkg/sql.(*planner).makeOptimizerPlan(0xc005efd450, 0x39fc560, 0xc00644c2a0, 0x32b8e38, 0x19, 0xbe52b8db5, 0x391d63e)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan_opt.go:162 +0x18b fp=0xc006434a18 sp=0xc006434990 pc=0x209988b
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).makeExecPlan(0xc005efd100, 0x39fc560, 0xc00644c2a0, 0xc005efd450, 0x6, 0x70b01f)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:981 +0x131 fp=0xc006434ad8 sp=0xc006434a18 pc=0x1fc00e1
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc005efd100, 0x39fc560, 0xc00644c2a0, 0xc005efd450, 0x7fc652d8a340, 0xc0059c5d90, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:868 +0x148 fp=0xc006434bd8 sp=0xc006434ad8 pc=0x1fbf3c8
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc005efd100, 0x39fc560, 0xc00644c2a0, 0x3a020a0, 0xc0062e5d80, 0xc00638e023, 0x6e, 0x2, 0xc0064fcd80, 0x7, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:459 +0xdf6 fp=0xc0064353b8 sp=0xc006434bd8 pc=0x1fbb976
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc005efd100, 0x39fc560, 0xc00644c2a0, 0x3a020a0, 0xc0062e5d80, 0xc00638e023, 0x6e, 0x2, 0xc0064fcd80, 0x7, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:101 +0x610 fp=0xc006435660 sp=0xc0064353b8 pc=0x1fba700
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc005efd100, 0x39fc4a0, 0xc0062e5c80, 0xc000959538, 0x5400, 0x15000, 0xc0009595d0, 0xc0063604b0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1237 +0x1418 fp=0xc006435ec8 sp=0xc006435660 pc=0x1fb04d8
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc000c54000, 0x39fc4a0, 0xc0062e5c80, 0xc005efd100, 0x5400, 0x15000, 0xc0009595d0, 0xc0063604b0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:428 +0xce fp=0xc006435f28 sp=0xc006435ec8 pc=0x1fabe2e
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl.func4(0xc000c54000, 0x39fc4a0, 0xc0062e5c80, 0xc000012400, 0xc00636aba4, 0xc0063604b0, 0xc005efd100, 0x5400, 0x15000, 0xc0009595d0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:338 +0xe6 fp=0xc006435f88 sp=0xc006435f28 pc=0x22323b6
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1333 +0x1 fp=0xc006435f90 sp=0xc006435f88 pc=0x75bd81
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:321 +0x1080

```

cc @justinj

--- a/pkg/kv/txn_coord_sender.go
+++ b/pkg/kv/txn_coord_sender.go
@@ -631,6 +631,23 @@ func (tc *TxnCoordSender) AugmentMeta(ctx context.Context, meta roachpb.TxnCoord
 	if tc.mu.txn.ID != meta.Txn.ID {
 		return
 	}
+
+	// If the TxnCoordMeta is telling us the transaction has been aborted, it's
+	// better if we don't ingest it. Ingesting it would possibly put us in an
+	// inconsistent state, with an ABORTED proto but with the heartbeat loop still
+	// running. When this TxnCoordMeta was received from DistSQL, it presumably
+	// followed a TxnAbortedError that was also received. If that error was also
+	// passed to us, then we've already aborted the txn and ingesting the meta
+	// would be OK. However, as it stands, if the TxnAbortedError followed a
+	// non-retriable error, than we don't get the aborted error (in fact, we don't
+	// get either of the errors; the client is responsible for rolling back).
+	// TODO(andrei): A better design would be to abort the txn as soon as any
+	// error is received from DistSQL, which would eliminate qualms about what
+	// error comes first.
+	if meta.Txn.Status != roachpb.PENDING {
+		return
+	}
+
 	tc.augmentMetaLocked(ctx, meta)
 }
 

--- a/pkg/kv/txn_coord_sender_test.go
+++ b/pkg/kv/txn_coord_sender_test.go
@@ -2384,3 +2384,23 @@ func TestLeafTxnClientRejectError(t *testing.T) {
 		t.Fatalf("expected UnhandledRetryableError(TransactionAbortedError), got: (%T) %v", err, err)
 	}
 }
+
+// Check that ingesting an Aborted txn record is a no-op. The TxnCoordSender is
+// supposed to reject such updates because they risk putting it into an
+// inconsistent state. See comments in TxnCoordSender.AugmentMeta().
+func TestAugmentTxnCoordMetaAbortedTxn(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	s := createTestDBWithContextAndKnobs(t, client.DefaultDBContext(), nil /* knobs */)
+	defer s.Stop()
+	ctx := context.Background()
+
+	txn := client.NewTxn(ctx, s.DB, 0 /* gatewayNodeID */, client.RootTxn)
+	txnProto := txn.Serialize()
+	txnProto.Status = roachpb.ABORTED
+	txn.AugmentTxnCoordMeta(ctx, roachpb.TxnCoordMeta{Txn: *txnProto})
+	// Check that the transaction was not updated.
+	txnProto = txn.Serialize()
+	if txnProto.Status != roachpb.PENDING {
+		t.Fatalf("expected PENDING txn, got: %s", txnProto.Status)
+	}
+}

